File: backend-node/src/routes/testcases.js
import { Router } from 'express';
import { v4 as uuidv4 } from 'uuid';
import { db } from '../db/database.js';
import { runGenerateTestCases } from '../services/jobs.js';

const router = Router();

router.post('/generate', async (req, res) => {
  const { requirement_id, format = 'BDD', count_hint = 'auto', additional_context = '' } = req.body;
  if (!requirement_id) return res.status(400).json({ detail: 'requirement_id is required' });

  const req_ = db.prepare('SELECT * FROM requirements WHERE id=?').get(requirement_id);
  if (!req_) return res.status(404).json({ detail: 'Requirement not found' });

  const jobId = uuidv4();
  db.prepare(`
    INSERT INTO generation_jobs (id, requirement_id, status, format, count_hint)
    VALUES (?,'?','PENDING',?,?)
  `.replace("'?'", '?')).run(jobId, requirement_id, format, count_hint);

  const job = db.prepare('SELECT * FROM generation_jobs WHERE id=?').get(jobId);
  res.status(202).json(job);

  // Fire-and-forget
  setImmediate(() => runGenerateTestCases({
    jobId, requirementId: requirement_id, projectId: req_.project_id,
    fmt: format, countHint: count_hint, additionalContext: additional_context,
  }));
});

router.get('/requirement/:requirementId', (req, res) => {
  const fmt = req.query.format;
  let sql = 'SELECT * FROM test_cases WHERE requirement_id=?';
  const params = [req.params.requirementId];
  if (fmt) { sql += ' AND format=?'; params.push(fmt.toUpperCase()); }
  sql += ' ORDER BY created_at';
  res.json(db.prepare(sql).all(...params).map(tc => ({ ...tc, edited: !!tc.edited })));
});

router.get('/:id', (req, res) => {
  const tc = db.prepare('SELECT * FROM test_cases WHERE id=?').get(req.params.id);
  if (!tc) return res.status(404).json({ detail: 'TestCase not found' });
  res.json({ ...tc, edited: !!tc.edited });
});

router.patch('/:id', (req, res) => {
  const tc = db.prepare('SELECT * FROM test_cases WHERE id=?').get(req.params.id);
  if (!tc) return res.status(404).json({ detail: 'TestCase not found' });
  const { title, content, priority, tags, scenario_type } = req.body;
  db.prepare(`
    UPDATE test_cases SET title=?, content=?, priority=?, tags=?, scenario_type=?, edited=1, updated_at=?
    WHERE id=?
  `).run(
    title ?? tc.title, content ?? tc.content, priority ?? tc.priority,
    tags ?? tc.tags, scenario_type ?? tc.scenario_type,
    new Date().toISOString(), req.params.id,
  );
  const updated = db.prepare('SELECT * FROM test_cases WHERE id=?').get(req.params.id);
  res.json({ ...updated, edited: !!updated.edited });
});

router.delete('/:id', (req, res) => {
  if (!db.prepare('SELECT id FROM test_cases WHERE id=?').get(req.params.id))
    return res.status(404).json({ detail: 'TestCase not found' });
  db.prepare('DELETE FROM test_cases WHERE id=?').run(req.params.id);
  res.status(204).end();
});

router.post('/export', (req, res) => {
  const { project_id, requirement_id, format = 'BDD', file_type = 'csv' } = req.body;
  let cases;
  if (requirement_id) {
    let sql = 'SELECT * FROM test_cases WHERE requirement_id=?';
    const p = [requirement_id];
    if (format !== 'BOTH') { sql += ' AND format=?'; p.push(format); }
    cases = db.prepare(sql + ' ORDER BY created_at').all(...p);
  } else if (project_id) {
    let sql = 'SELECT tc.* FROM test_cases tc JOIN requirements r ON tc.requirement_id=r.id WHERE r.project_id=?';
    const p = [project_id];
    if (format !== 'BOTH') { sql += ' AND tc.format=?'; p.push(format); }
    cases = db.prepare(sql + ' ORDER BY tc.created_at').all(...p);
  } else {
    return res.status(400).json({ detail: 'Provide project_id or requirement_id' });
  }
  if (!cases.length) return res.status(404).json({ detail: 'No test cases found' });

  const ts = new Date().toISOString().replace(/[:.]/g, '-').slice(0, 19);
  if (file_type === 'json') {
    res.setHeader('Content-Disposition', `attachment; filename="testcases_${ts}.json"`);
    return res.json(cases.map(tc => ({ ...tc, edited: !!tc.edited })));
  }
  const header = 'ID,Title,Format,Priority,Scenario Type,Tags,Content,Edited,Created At\n';
  const rows   = cases.map(tc =>
    [tc.id, tc.title, tc.format, tc.priority, tc.scenario_type || '', tc.tags || '',
     `"${(tc.content || '').replace(/"/g, '""')}"`, tc.edited ? 'true' : 'false', tc.created_at]
    .join(',')
  ).join('\n');
  res.setHeader('Content-Disposition', `attachment; filename="testcases_${ts}.csv"`);
  res.setHeader('Content-Type', 'text/csv; charset=utf-8');
  res.send('\uFEFF' + header + rows);   // BOM for Excel compatibility
});

export default router;

File: backend/app/api/routes/testcases.py
"""Test case CRUD, generation trigger, and export."""
from __future__ import annotations

import asyncio
import csv
import io
import json
from datetime import datetime

from fastapi import APIRouter, Depends, HTTPException, Query
from fastapi.responses import StreamingResponse
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_job_repo, get_requirement_repo, get_testcase_repo
from app.core.exceptions import NotFoundError
from app.db.repository import JobRepository, RequirementRepository, TestCaseRepository
from app.db.session import get_db
from app.schemas.schemas import ExportRequest, GenerateRequest, JobRead, TestCaseCreate, TestCaseRead, TestCaseUpdate

router = APIRouter()


@router.post("", response_model=TestCaseRead, status_code=201)
async def create_test_case(
    payload: TestCaseCreate,
    req_repo: RequirementRepository = Depends(get_requirement_repo),
    tc_repo: TestCaseRepository = Depends(get_testcase_repo),
    db: AsyncSession = Depends(get_db),
):
    """Manually create a test case."""
    try:
        await req_repo.get_or_raise(payload.requirement_id)
    except NotFoundError as exc:
        raise HTTPException(404, str(exc))

    tc = await tc_repo.create(
        requirement_id=payload.requirement_id,
        title=payload.title,
        format=payload.format.upper(),
        content=payload.content,
        priority=payload.priority.upper(),
        tags=payload.tags,
        scenario_type=payload.scenario_type,
        edited=True,
    )
    await db.commit()
    await db.refresh(tc)
    return tc


@router.post("/generate", response_model=JobRead, status_code=202)
async def trigger_generation(
    payload: GenerateRequest,
    req_repo: RequirementRepository = Depends(get_requirement_repo),
    job_repo: JobRepository = Depends(get_job_repo),
    db: AsyncSession = Depends(get_db),
):
    """Start async test case generation. Returns a job to poll/stream."""
    try:
        req = await req_repo.get_or_raise(payload.requirement_id)
    except NotFoundError as exc:
        raise HTTPException(404, str(exc))

    job = await job_repo.create(
        requirement_id=payload.requirement_id,
        format=payload.format,
        count_hint=payload.count_hint,
        status="PENDING",
    )
    await db.commit()
    await db.refresh(job)

    from app.workers.tasks import run_generate_test_cases
    asyncio.create_task(run_generate_test_cases(
        job_id=job.id,
        requirement_id=payload.requirement_id,
        project_id=req.project_id,
        fmt=payload.format,
        count_hint=payload.count_hint,
        additional_context=payload.additional_context or "",
    ))

    return job


@router.get("/requirement/{requirement_id}", response_model=list[TestCaseRead])
async def list_test_cases(
    requirement_id: str,
    format: str | None = Query(None),
    tc_repo: TestCaseRepository = Depends(get_testcase_repo),
):
    return await tc_repo.list_by_requirement(requirement_id, fmt=format)


@router.get("/{testcase_id}", response_model=TestCaseRead)
async def get_test_case(
    testcase_id: str,
    tc_repo: TestCaseRepository = Depends(get_testcase_repo),
):
    try:
        return await tc_repo.get_or_raise(testcase_id)
    except NotFoundError as exc:
        raise HTTPException(404, str(exc))


@router.patch("/{testcase_id}", response_model=TestCaseRead)
async def update_test_case(
    testcase_id: str,
    payload: TestCaseUpdate,
    tc_repo: TestCaseRepository = Depends(get_testcase_repo),
    db: AsyncSession = Depends(get_db),
):
    try:
        tc = await tc_repo.get_or_raise(testcase_id)
    except NotFoundError as exc:
        raise HTTPException(404, str(exc))

    for k, v in payload.model_dump(exclude_none=True).items():
        setattr(tc, k, v)
    tc.edited     = True
    tc.updated_at = datetime.utcnow()
    await db.commit()
    await db.refresh(tc)
    return tc


@router.delete("/{testcase_id}", status_code=204)
async def delete_test_case(
    testcase_id: str,
    tc_repo: TestCaseRepository = Depends(get_testcase_repo),
    db: AsyncSession = Depends(get_db),
):
    try:
        tc = await tc_repo.get_or_raise(testcase_id)
    except NotFoundError as exc:
        raise HTTPException(404, str(exc))
    await tc_repo.delete(tc)
    await db.commit()


@router.post("/export")
async def export_test_cases(
    payload: ExportRequest,
    tc_repo: TestCaseRepository = Depends(get_testcase_repo),
):
    """Export test cases as CSV or JSON."""
    fmt_filter = None if payload.format == "BOTH" else payload.format

    if payload.requirement_id:
        cases = list(await tc_repo.list_by_requirement(payload.requirement_id, fmt=fmt_filter))
    elif payload.project_id:
        cases = list(await tc_repo.list_by_project(payload.project_id, fmt=fmt_filter))
    else:
        raise HTTPException(400, "Provide either requirement_id or project_id")

    if not cases:
        raise HTTPException(404, "No test cases found")

    ts = datetime.utcnow().strftime("%Y%m%d_%H%M%S")

    if payload.file_type == "json":
        body = json.dumps([
            {
                "id": tc.id, "title": tc.title, "format": tc.format,
                "priority": tc.priority, "scenario_type": tc.scenario_type,
                "tags": tc.tags, "content": tc.content, "edited": tc.edited,
                "created_at": tc.created_at.isoformat(),
            }
            for tc in cases
        ], indent=2, ensure_ascii=False).encode()
        return StreamingResponse(
            io.BytesIO(body), media_type="application/json",
            headers={"Content-Disposition": f'attachment; filename="testcases_{ts}.json"'},
        )

    buf = io.StringIO()
    w   = csv.writer(buf)
    w.writerow(["ID", "Title", "Format", "Priority", "Scenario Type", "Tags", "Content", "Edited", "Created At"])
    for tc in cases:
        w.writerow([tc.id, tc.title, tc.format, tc.priority, tc.scenario_type or "",
                    tc.tags or "", tc.content, tc.edited, tc.created_at.isoformat()])
    buf.seek(0)
    return StreamingResponse(
        io.BytesIO(buf.getvalue().encode("utf-8-sig")), media_type="text/csv",
        headers={"Content-Disposition": f'attachment; filename="testcases_{ts}.csv"'},
    )
